wallet/src/account/transaction_list/mod.rs

```diff
@@ -0,0 +1,249 @@
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct TransactionInfo {
+    pub txid: Id<Transaction>,
+    pub tx_type: TxType,
+    pub timestamp: Option<BlockTimestamp>,
+    pub state: TxState,
+}
+
+#[derive(Debug, Clone)]
+pub struct TransactionList {
+    /// How many transactions are in the single page (currently it's always 10)
+    pub count: usize,
+
+    /// How many transactions the user has seen and scrolled through (normally it can be 0, 10, 20, etc.)
+    pub skip: usize,
+
+    /// Total number of transactions
+    pub total: usize,
+
+    /// Transaction list from the selected page
+    pub txs: Vec<TransactionInfo>,
+}
+
+struct TxRef<'a> {
+    block_height: Option<BlockHeight>,
+    tx_data: &'a TxData,
+}
+
+// Most recent transactions go first
+fn compare_tx_ref(a: &TxRef, b: &TxRef) -> Ordering {
+    match (a.block_height, b.block_height) {
+        (None, None) => Ordering::Equal,
+        (None, Some(_)) => Ordering::Less,
+        (Some(_), None) => Ordering::Greater,
+        (Some(a), Some(b)) => b.cmp(&a),
+    }
+}
+
+fn own_output(key_chain: &AccountKeyChain, output: &TxOutput) -> bool {
+    match output {
+        TxOutput::Transfer(_, dest) | TxOutput::LockThenTransfer(_, dest, _) => KeyPurpose::ALL
+            .iter()
+            .any(|purpose| key_chain.get_leaf_key_chain(*purpose).is_destination_mine(dest)),
+        TxOutput::Burn(_) => false,
+        TxOutput::CreateStakePool(_, _) => false,
+        TxOutput::ProduceBlockFromStake(_, _) => false,
+        TxOutput::CreateDelegationId(_, _) => false,
+        TxOutput::DelegateStaking(_, _) => false,
+    }
+}
+
+fn own_input<'a>(
+    key_chain: &AccountKeyChain,
+    output_cache: &'a OutputCache,
+    input: &TxInput,
+) -> Option<&'a TxOutput> {
+    match input {
+        TxInput::Utxo(utxo) => match output_cache.txs_with_unconfirmed().get(&utxo.tx_id()) {
+            Some(tx) => tx
+                .outputs()
+                .get(utxo.output_index() as usize)
+                .filter(|&output| own_output(key_chain, output)),
+            None => None,
+        },
+        TxInput::Account(_) => None,
+    }
+}
+
+fn get_transaction(
+    key_chain: &AccountKeyChain,
+    output_cache: &OutputCache,
+    tx_data: &TxData,
+) -> WalletResult<TransactionInfo> {
+    let timestamp = tx_data.state().timestamp();
+
+    let all_inputs = tx_data.get_transaction().inputs();
+    let all_outputs = tx_data.get_transaction().outputs();
+
+    let own_inputs = all_inputs
+        .iter()
+        .filter_map(|input| own_input(key_chain, output_cache, input))
+        .collect::<Vec<_>>();
+
+    let own_outputs = tx_data
+        .get_transaction()
+        .outputs()
+        .iter()
+        .filter(|output| own_output(key_chain, output))
+        .collect::<Vec<_>>();
+
+    let non_own_outputs = tx_data
+        .get_transaction()
+        .outputs()
+        .iter()
+        .filter(|output| !own_output(key_chain, output));
```

Comment:
I know it's an optimization, but should `own_outputs` and `non_own_outputs` be combined in a simple for loop so we're not looping over outputs twice (if it could be a large number of transactions)

Reply:
Right, thanks. I did not find out how to do it last time. I found it now with the help of ChatGPT.